datajoint/autopopulate.py

```diff
@@ -45,35 +45,48 @@ def populate(self, restriction=None, suppress_errors=False, reserve_jobs=False):
         :param restriction: restriction on rel.populate_relation - target
         :param suppress_errors: suppresses error if true
         :param reserve_jobs: currently not implemented
+        :param max_attempts: maximal number of times a TransactionError is caught before populate gives up
         """
-        assert not reserve_jobs, NotImplemented  # issue #5
 
+        assert not reserve_jobs, NotImplemented   # issue #5
         error_list = [] if suppress_errors else None
-
         if not isinstance(self.populate_relation, RelationalOperand):
             raise DataJointError('Invalid populate_relation value')
 
-        unpopulated = (self.populate_relation - self.target) & restriction
+        self.conn.cancel_transaction()  # rollback previous transaction, if any
+
+        if not isinstance(self, Relation):
+            raise DataJointError('Autopopulate is a mixin for Relation and must therefore subclass Relation')
 
+        unpopulated = (self.populate_relation - self.target) & restriction
         for key in unpopulated.project():
-            try:
-                while True:
-                    try:
-                        with self.conn.transaction():
-                            if not key in self.target:  # key not populated yet
-                                logger.info('Populating: ' + str(key))
-                                self._make_tuples(dict(key))
-                        break
-                    except TransactionError as tr_err:
-                        if suppress_errors:
-                            error_list.append((key,tr_err))
-                        tr_err.resolve()
-                        logger.info('Resolved transaction error raised by {0:s}.'.format(tr_err.culprit))
-            except Exception as error:
-                if not suppress_errors:
-                    raise
+            self.conn.start_transaction()
+            if key in self.target:  # already populated
+                self.conn.cancel_transaction()
+            else:
+                logger.info('Populating: ' + str(key))
+                try:
+                    for attempts in range(max_attempts):
+                        try:
+                            self._make_tuples(dict(key))
+                            break
+                        except TransactionError as tr_err:
+                            self.conn.cancel_transaction()
+                            tr_err.resolve()
+                            self.conn.start_transaction()
+                            logger.info('Transaction error in {0:s}.'.format(tr_err.culprit))
+                    else:
+                        raise DataJointError(
```

Comment:
The attempts are not always multiple, so I would just say `%s._make_tuples transaction failed.`